### ISSUE TITLE
Bug: Logging out using redirect authentication does not log out the user

### ISSUE THREAD

**Issue by cfg-fadica** [wpengine/faustjs#1147]
When logging out using the redirect method of authentication, the frontend cookie persists and the user remains logged in to the application. I was able to reproduce this on a fresh install.

## Applicable Versions

- `@faustjs/core` version: 0.15.7
- `@faustjs/react` version: 0.15.7
- `@faustjs/next` version: 0.15.9
- WordPress version: 6.1

## Steps To Reproduce

1. Install my repo https://github.com/cfg-fadica/faustjs-logout-demo
2. Connect the repo to a WordPress backend
3. Start the server and visit localhost:3000/authpage
4. Click log in and log in with valid credentials
5. Once redirected back to the frontend app, click the log out link. This will show you the "unauthenticated content" message.
6. Visit localhost:3000/authpage again and you will see the authenticated content and the log out link.

Contents of src/pages/authpage.tsx
```TypeScript
import {client} from 'client'

export default function MyPage() {
    const { isLoading: isAuthLoading, isAuthenticated } = client.auth.useAuth({
        shouldRedirect: true,
    })

    const { isLoggedOut, logout } = client.auth.useLogout()


    /**
     * Not authenticated
     */
    if ( !isAuthenticated || isLoggedOut ) {
        return <div>Not authenticated</div>
    }

    return (
        <div>
            <p>My auth content</p>
            <a href="#" onClick={() => logout()}>Log out</a>
        </div>
    )
}
```

Link to code example: https://github.com/cfg-fadica/faustjs-logout-demo

## The current behavior
User remains logged in after using the `logout()` function exported from `client.auth.useLogout()`. The frontend app cookie persists.

## The expected behavior
When the `logout()` function is fired, frontend app cookie would be destroyed and user would be logged out, unable to access the authenticated content without logging in again.

**Comment by theodesp** [comment]
Hey @cfg-fadica. Thank you for mentioning this one. I think this is the expected result when using this hook and providing the `shouldRedirect: true,`.

You see according to the [docs here](https://legacy.faustjs.org/docs/next/reference/hooks/useAuth) when you use this hook into your page and provide the `shouldRedirect: true` it will automatically try to login with your WordPress instance. Since you've been already logged in with WordPress, then this process will happen again once you reload the page.

If you logout from WordPress then the whole process is reversed and you will have to login again.

If you wish to disable redirects on load you should use the `shouldRedirect: false` so in that case when you re-load the page then you are still not authenticated:

<img width="355" alt="Screenshot 2022-11-16 at 10 59 52" src="https://user-images.githubusercontent.com/328805/202163145-a4ca0b0a-c263-4ecb-81ea-168713d39e6e.png">

Now if you wish to control this  better you should be using a login button instead of having `useAuth` here and try to stop the redirect from happening.

I hope this paints a better picture.

**Comment by cfg-fadica** [comment]
Hi @theodesp,
Thank you for your response. I am confused by your response, but what I think I have determined is that I think the bug I am reporting is `logout()` should destroy the WordPress cookies in a redirect authentication model.

 Do you mean that I have to implement my own separate log out from WordPress in order for this to work? I thought that `logout()` would do that too. I am really confused as to what the `logout()` function actually accomplishes in this case. 

I've used `logout()` successfully with local authentication, but not with redirect authentication. I have a case where I need to switch my auth type for my application. To me it would seem that `logout()` would destroy the WordPress cookie(s), thus redirecting a user to the log in page correctly.

So, in conclusion, I am reporting that the actual bug is that WordPress cookies are not destroyed.

**Comment by jonjakoblich** [comment]
Hi there. I had to switch accounts. I am the ghost above, formerly cfg-fadica.

**Comment by theodesp** [comment]
@jonjakoblich When you call logout, the framework invalidates the refresh [token here](https://github.com/wpengine/faustjs/blob/canary/packages/core/src/server/auth/middleware.ts#L100-L116) and clears the cookie. 

You can verify this by loading the page using `shouldRedirect: true,` so that the cookie is populated:
<img width="1790" alt="Screenshot 2022-11-17 at 10 46 53" src="https://user-images.githubusercontent.com/328805/202426766-35e229c3-bc50-43f1-800a-d875361b55b6.png">


Then go to your code and change `shouldRedirect: false,` and then try to logout. You will see that after that there will be no cookie in the page:
<img width="1792" alt="Screenshot 2022-11-17 at 10 46 40" src="https://user-images.githubusercontent.com/328805/202426466-c44eecfa-7dc8-4ab6-8d53-7862cc725af2.png">

If you reload the page with `shouldRedirect: false` then you still be not authenticated. The tricky part to understand is that when you use this code in the page:

```
const { isLoading: isAuthLoading, isAuthenticated } = client.auth.useAuth({
        shouldRedirect: true,
    })
```
It will try to authenticate, get the access token and populate the auth cookie every time. When you click logout  it will temporarily clear the cookie for a moment but then the whole process will start again as the ` shouldRedirect: true,`will be still there.

Check this recording:

https://user-images.githubusercontent.com/328805/202427951-22f09b6a-10c7-4b11-a106-f9523e7a6fd5.mov

**Comment by jonjakoblich** [comment]
Hi @theodesp. Thank you for this. I replicated the same steps you took using `shouldRedirect: true` and captured it on video.


https://user-images.githubusercontent.com/106980918/202490802-23b16619-5f4e-4793-9818-abc68eecbe29.mp4


Faust does not terminate the WordPress session, and to me it seems that it should have been terminated on `logout()`. Therefore, when the page redirects it sees that the user is still authenticated in WordPress and then generates a valid refresh token giving the appearance that the user never logged out from the frontend app.  This seems like a bug. 

What is the effective way to log out a user using `shouldRedirect: true`?

**Comment by theodesp** [comment]
Hey @jonjakoblich. Thats is true. We don't destroy the session there. Maybe this is a good feature request that we may investigate. I will create a ticket in our backlog.

As for using the `shouldRedirect: true` I think the best way is to have a button for specifically login once unauthenticated and use a combination of `useState` and `useEffect`. Here how it should look like roughly:

```jsx
import * as React from 'react';
import {client} from 'client'

export default function MyPage() {
    // start with no redirect
    const [shouldRedirect, setShouldRedirect] = React.useState(false);
    const { isLoading: isAuthLoading, isAuthenticated } = client.auth.useAuth({
        shouldRedirect
    });

    const login = () => {
        // redirect to login once user clicks button.
        setShouldRedirect(true);
    };

    React.useEffect(() => {
        // Once authenticated then disable the automatic redirect.
        if (isAuthenticated && shouldRedirect) {
            setShouldRedirect(false);
        }
    }, [shouldRedirect, isAuthenticated])

    const { isLoggedOut, logout } = client.auth.useLogout()

    /**
     * Not authenticated
     */
    if ( !isAuthenticated || isLoggedOut ) {
        return <div>Not authenticated. <a href="#" onClick={() => login()}>Login</a></div>
    }

    return (
        <div>
            <p>My auth content</p>
            
            <a href="#" onClick={() => logout()}>Log out</a>
        </div>
    )
}
```

I've added a recording. for this functionality:


https://user-images.githubusercontent.com/328805/202502420-5f32fe78-a451-40b7-84c4-860e64912c58.mov

**Comment by theodesp** [comment]
Sure I've added a ticket for this and will follow up soon.

**Comment by JEverhart383** [comment]
@Fran-A-Dev and I experienced this behavior as well when using redirect-based auth to answer a question in Discord as an FYI

**Comment by phil-sola** [comment]
Just chiming in here as well to say it would be great to see this feature added to the framework.

I’ve just had this issue today when messaging on Discord. It’s pretty confusing, especially for customers on the front end who won’t have access to the backend to logout in both places. 

Would love to see this behaviour added in without workarounds 😀

**Comment by akalex-x** [comment]
Would love to see this implemented as well!
